### ISSUE TITLE
homme with fpmodel strict crashes for omp runs

### ISSUE THREAD

**Issue by oksanaguba** [E3SM-Project/E3SM#2515]
Low priority.

When homme is built with fpmodel strict on skybridge (intel 17,18) or anvil (intel 17), one of baro2c runs crashes. We reduced a namelist, removed restart, etc. The list that still shows the issue is below. Commands are below. valgrind did not help (too slow to get past init), totalview did not show anything. Another route is to try gnu on anvil (on skybridge netcdf is not built for gnu) with all possible debug/check flags or to build compilers, etc. very bare, no vectorization, to make valgrind work (AMB expertise).

The crash is usually in repro_sum, but nondeterministic, so, printstatements do not help (crash either does not occur or occurs for repro_sum of a different variable each time).

```
export OMP_NUM_THREADS=4
mpiexec -n 4 ../../test_execs/baroC/baroC < input.nl
#ttoalview on skybridge
mpiexec -tv -n 4 ../../test_execs/baroC/baroC < input.nl
```

```
&ctl_nl
NThreads      = 4
partmethod    = 4
topology      = "cube"
test_case     = "jw_baroclinic"
ne            = 2
qsize = 1
nmax          = 3
statefreq     = 1
restartfreq   = -100
!restartfile   = "./restart/R000000864"
!runtype = 2
tstep         = 30
tstep_type    = 5
integration   = "explicit"
smooth        = 0.0000
nu            = 4e16
nu_s          = 4e16
nu_p          = 0
nu_top = 0
hypervis_order = 2
hypervis_subcycle = 1
u_perturb      = 1
initial_total_mass = 0
limiter_option = 8
/
&solver_nl
precon_method = "identity"
maxits        = 50
tol           = 1.e-7
/
&filter_nl
filter_type   = "taylor"
transfer_type = "bv"
filter_freq   = 0
filter_mu     = 0.25D0
p_bv          = 12.0D0
s_bv          = .666666666666666666D0
wght_fm       = 0.10D0
kcut_fm       = 2
/
&vert_nl
vform         = "ccm"
vfile_mid     = "vcoord/camm-26.ascii"
vfile_int     = "vcoord/cami-26.ascii"
/
&analysis_nl
output_prefix    = "baro2c-run2-omp-"
output_timeunits = 0
output_frequency = 3
output_varnames1 = 'u', 'v', 'ps', 'T'
interp_type      = 0          ! native high order
output_type      = 'netcdf'
io_stride        = 8
/
```

**Comment by ambrad** [comment]
HORIZ_OPENP is on, COLUMN_NP is off, right?

**Comment by ambrad** [comment]
Also, IIRC, manifests with 1 rank, 4 threads? I.e., 4 ranks not needed?

**Comment by oksanaguba** [comment]
on 1 rank nothing crashes, so, 4r+4t are needed, i believe. and yes, hor, threading on, vert. off. thanks!

**Comment by ambrad** [comment]
One possibility, or else I'm confused (easily possible): Is `accum%IEner_wet` every set anywhere?

**Comment by ambrad** [comment]
Oh, yes, in prim_energy_halftimes. More digging.

**Comment by ambrad** [comment]
After some pair programming with Oksana, I think we got it. In wrap_repro_sum, the NaN check iterates i = 1, nsize_use, where in the failing case nsize_use = nelemd. Thus, each thread checks values for all ie. The loop should actually be i = nets, nete, or it should occur after the barrier. In its current state, a thread checks a value before it might have been written. At startup, the value might be uninitialized, for example.

**Comment by oksanaguba** [comment]
forgot to add this to the description first time: ran the code 100 times to check this mod:
```
(for i in $(seq 100) ;do
echo ">>> $i"
mpiexec -n 4 ../../test_execs/baroC/baroC < input.nl ; done ) |& grep "NaN \|>>>"
```

**Comment by oksanaguba** [comment]
should be fixed by #2521
